doc/authentication-proposal.md

```diff
@@ -0,0 +1,114 @@
+Authenticating with Hosted Pub Repositories Proposal
+====================================================
+
+This document specifies how the REST API could be authenticated with the pub CLI
+tool.
+
+This proposal is mostly based on
+[RFC7235](https://datatracker.ietf.org/doc/html/rfc7235) - HTTP Authentication
+specifications. This ensures exists un-protected endpoints could be protected
+just by setting up reverse proxies like NGINX, Apache, Traefik, which already
+supports this specifications.
+
+## Requesting authentication
+
+If the repository requires special authentication to access resources or upload
+new packages, it has to respond with `401 Unauthorized` status code with
+`WWW-Authenticate` header. This header should specify authentication method that
+should be used to gain access to the resource.
+
+WWW-Authenticate header syntax:
+
+```plain
+WWW-Authenticate: <type> [realm=<realm>][, charset="UTF-8"]
+```
+
+> `realm` parameter is completely optional, and is not used in this proposal.
+> You can read more about this parameter here:
+> https://datatracker.ietf.org/doc/html/rfc7235#section-2.2
+
+Pub CLI will only support **Basic** and **Bearer** authentication methods by
+default.
+
+## Authentication flow
+
+After receiving `WWW-Authenticate` header, pub CLI will prompt user for
+the credentials like this:
+
+```plain
+user@machine$ pub get
+Please enter required credentials to authenticate with "https://myserver.com"
+hosted repository.
+
+Username: bob
+Password: password
+
+Please enter required credentials to authenticate with "https://pub.example.com"
+hosted repository.
+
+Bearer token: 8O868XsPJm-F5nyEzXfa9-YWFvrd3O8r
+```
+
+After providing credentials, the client will send those credentials to the
+server in `Authorization` header.
+
+
+```dart
+// Bearer authentication
+{ 'Autorization': 'Bearer $token' }
+
+// Basic authentication
+{ 'Autorization': 'Basic ' + base64('$username:$password') }
+```
+
+### Explicit authentication
+
+Users can also explicitly define authentication credentials even before server
+asks for it using a modified **pub login** command. By default if no argument is
+presented, **pub login** will behave as previously: authenticate with Google.
+But if you you specify server to login, it will try to authenticate with the
+given server instead.
+
+```plain
+pub login https://myspuberver.dev
+```
+
+To discover authentication method, client will send **GET** request to `/login`
```

Comment:
But what if `/publish` endpoint utilizes some sort of resources? Everytime we send request to that endpoint, we can create some garbage for the repository.

For example a server might add an entry to the "UploadRequests" table on every request which might be used to generate unique URL for the upload using models auto increment ID.

Reply:
hmm, what if there was a `/server-info.json` file, and it contains the information (or something like this).

It's easy to write a JSON file to an S3 bucket (or GCS bucket)... that could point contain how to authenticate and how to publish in a programmatic form.

I imagine that if you're writing a custom server it might be attractive to simply use an S3 bucket, and point the `dart pub` client at an S3 bucket. This is not fully conceived yet. But adding a `/login` end-point would make this harder.

--------

I need to find a little time to figure this out..